tpls/desul/include/desul/atomics/Compare_Exchange_OpenACC.hpp

```diff
@@ -0,0 +1,269 @@
+  return reinterpret_cast<T&>(return_val);
+}
+
+#pragma acc routine seq
+template <class T, class MemoryOrder, class MemoryScope>
+std::enable_if_t<std::is_floating_point<T>::value && (sizeof(T) == 4), T> device_atomic_compare_exchange(T* const dest, T compare, T value, MemoryOrder, MemoryScope) {
+  static_assert(sizeof(float) == 4,
+                "this function assumes an float is 32-bit");
+  float return_val = atomicCAS(reinterpret_cast<float*>(dest),
+                                      reinterpret_cast<float&>(compare),
+                                      reinterpret_cast<float&>(value));
+  return reinterpret_cast<T&>(return_val);
+}
+
+#ifndef DESUL_CUDA_ARCH_IS_PRE_PASCAL
+
+#pragma acc routine seq
+template <class T, class MemoryOrder, class MemoryScope>
+std::enable_if_t<std::is_floating_point<T>::value && (sizeof(T) == 8), T> device_atomic_compare_exchange(T* const dest, T compare, T value, MemoryOrder, MemoryScope) {
+  static_assert(sizeof(double) == 8,
+                "this function assumes an double is 64-bit");
+  double return_val = atomicCAS(reinterpret_cast<double*>(dest),
+                                      reinterpret_cast<double&>(compare),
+                                      reinterpret_cast<double&>(value));
+  return reinterpret_cast<T&>(return_val);
+}
+
+#else
+
+#pragma acc routine seq
+template <class T, class MemoryOrder, class MemoryScope>
+std::enable_if_t<std::is_floating_point<T>::value && (sizeof(T) == 8), T> device_atomic_compare_exchange(T* const dest, T compare, T value, MemoryOrder, MemoryScope scope) {
+  T current_val = *dest;
+  if (acc_on_device(acc_device_not_host)) {
+    printf("DESUL error in device_atomic_compare_exchange(): Not supported atomic "
+                  "operation in the OpenACC backend\n");
+  }
+  // Acquire a lock for the address
+  //while (!lock_address_openacc((void*)dest, scope)) {
+  //}
+  //device_atomic_thread_fence(MemoryOrderAcquire(), scope);
+  if (current_val == compare) {
+    *dest = value;
+    //device_atomic_thread_fence(MemoryOrderRelease(), scope);
+  }
+  //unlock_address_openacc((void*)dest, scope);
+  return current_val;
+}
+
+#endif
+
+#else
+
+#pragma acc routine seq
+template <class T, class MemoryOrder, class MemoryScope>
+T device_atomic_compare_exchange(T* dest, T compare, T value, MemoryOrder, MemoryScope scope) {
+  T current_val = *dest;
+  if (acc_on_device(acc_device_not_host)) {
+    printf("DESUL error in device_atomic_compare_exchange(): Not supported atomic "
+                  "operation in the OpenACC backend\n");
+  }
+  // Acquire a lock for the address
+  //while (!lock_address_openacc((void*)dest, scope)) {
+  //}
+  //device_atomic_thread_fence(MemoryOrderAcquire(), scope);
+  if (current_val == compare) {
+    *dest = value;
+    //device_atomic_thread_fence(MemoryOrderRelease(), scope);
+  }
+  //unlock_address_openacc((void*)dest, scope);
+  return current_val;
```

Comment:
Is that not something you can achieve with an acc atomic capture clause?

Reply:
The `acc atomic capture` can not handle "compare" part in the compare-exchange operation.